keps/sig-node/4112-passdown-resources-to-cri/README.md

```diff
@@ -0,0 +1,1300 @@
+-->
+
+The proposal only adds new informational data to the CRI API between kubelet
+and the container runtime with no user-visible changes which mitigates possible
+risks considerably.
+
+Data duplication/inconsistency with native resources could be considered a risk
+as those are passed down to CRI both as "raw" requests and limits and as
+"translated" resource control parameters (like cpu shares oom scoring etc). But
+this should be largely mitigated by code reviews and unit tests.
+
+## Design Details
+
+The proposal is that kubelet discloses full resources information from the
+PodSpec to the container runtime. This is accomplished by extending the
+ContainerConfig, UpdateContainerResourcesRequest and PodSandboxConfig messages
+of the CRI API.
+
+With this information, the runtime can for example do detailed resource
+allocation so that CPU, memory and other resources for each container are
+optimally aligned.
+
+The resource information is included in PodSandboxConfig so that the runtime
+can see the full picture of Pod's resource usage at Pod creation time, for
+example enabling more holistic resource allocation and thus better
+interoperability between containers inside the Pod.
+
+Also the CreateContainer request is extended to include the unmodified resource
+requirements. This make it possible for the CRI runtime to detect any changes
+in the pod resources that happen between the Pod creation and container
+creation in e.g.  scenarios where in-place pod updates are involved.
+
+The UpdatePodSandboxResources CRI message is also updated when/if that is
+introduced by the [KEP-1287][kep-1287] Beta ([PR][kep-1287-beta-pr]).
+
+Information about the Pod-level resources are added when/if the Pod-level
+resources enhancement [KEP-2837][kep-2837] Alpha ([PR][kep-2837-alpha-pr]) is
+implemented.
+
+### CRI API
+
+#### PodSandboxConfig
+
+The PodSandboxConfig message (part of the RunPodSandbox request) will be
+extended to contain information about resources of all its containers known at
+the pod creation time. The container runtime may use this information to make
+preparations for all upcoming containers of the pod. E.g. setup all needed
+resources for a VM-based pod or prepare for optimal allocation of resources of
+all the containers of the Pod. However, the container runtime may continue to
+operate as they did (before this enhancement). That is, it can ignore
+the resource information presented here and allocate resources for each
+container separately at container creation time with the `CreateContainer`
+request.
+
+```diff
+ message PodSandboxConfig {
+ 
+ ...
+ 
+     // Optional configurations specific to Linux hosts.
+     LinuxPodSandboxConfig linux = 8;
+     // Optional configurations specific to Windows hosts.
+     WindowsPodSandboxConfig windows = 9;
++
++    // Kubernetes resource spec of the containers in the pod.
++    PodResourceConfig pod_resources = 10;
```

Comment:
what runtime suppose to do when this does not match values provided in `resources` of `LinuxPodSandboxConfig`?

Reply:
There should be no overlapping information between these two structures. Thus a conflict should not be possible

Reply:
I am not sure I understand. Maybe I am missing something here, but you should be able to calculate what is in `resources` based on pod spec. Is it correct? So some implementation may take dependency on specific way one is calculated. Thus the question, what if they will not be matching at some point. Also if they are inter-dependent, perhaps some of it may be deprecated?

Reply:
The runtime is not calculating resources of `LinuxPodSandboxConfig`. They are set by the kubelet